### PR TITLE
[terra-clinical-result] Restructure the header name cell to be related

### PR DESCRIPTION
**Pull request by chenyixuan625**
### Summary

**What was changed:**
Add two optional props for the consumers to provide a meaningful string for the name or unit, instead of the displayed abbreviations.

**Why it was changed:**
The accessibility tooling can read out the full name of the header name and unit in the header name cell if needed.

### Testing

This change was tested using:
Jest, WDIO tests and Screen reader testing
Look at a ResultNameHeaderCell, and ensure the tooling announces the name and unit pair as related data(meaningful string would be read out instead of just abbreviations).

https://github.com/cerner/terra-clinical/assets/44185348/d77c7965-3423-4593-b6c7-2d7f1af56f72

- [x] WDIO
- [x] Jest
- [x] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

### Reviews

In addition to engineering reviews, this PR needs:

- [x] UX review
- [x] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-9076

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

**Review by vinaybhargavar — APPROVED** [review]

File: packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx [hunk 1]
```diff
@@ -17,6 +18,14 @@ const propTypes = {
    * Content to be displayed as the supporting Unit of Measure row label.
    */
   unit: PropTypes.string,
+  /**
+   * Full name of the Result Name for accessibility toolings.
+   */
+  fullResultName: PropTypes.string,
+  /**
+   * Full name of the unit for accessibility toolings.
```
Comment (kolkheang) [inline packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx:26]: Could you include "unit of measure" similar to the one above and with an example?

File: packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx [hunk 2]
```diff
@@ -17,6 +18,14 @@ const propTypes = {
    * Content to be displayed as the supporting Unit of Measure row label.
    */
   unit: PropTypes.string,
+  /**
+   * Full name of the Result Name for accessibility toolings.
```
[the remark this answers is not in the document]
Comment (chenyixuan625) [reply packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx]: Add an example: https://github.com/cerner/terra-clinical/pull/867/files#diff-10a8f15ae7fe78637c3fb3926670dd1d257ecc9c7a37f10b722d3ba6d579ea04R22

**Review by kolkheang — APPROVED** [review]

**Comment by chenyixuan625** [comment]
@sdadn / @sycombs May I get some eyes on this PR when you guys have a second? TIA!

**Review by brooksienoodlesoup — APPROVED** [review]
Tested in:

- Chrome 108.0.5359.94 (Official Build) (arm64) + VO (Ventura 13.2.1)
- Edge 114.0.1823.43 (Official build) (64-bit) + JAWS 2023.2303.144

Both VO and JAWS announced the expanded header text as expected.  Nice work!

*NOTE:* The JAWS Virtual PC Cursor (red visual focus indicator), as well as the VO black visual focus indicator did not move to cover the content that was being announced.  This is not a blocking error, but something that would be great to figure out how to work correctly in the future.

File: packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx [hunk 3]
```diff
@@ -59,16 +70,51 @@ const ResultNameHeaderCell = (props) => {
...
+          </div>
+          <VisuallyHiddenText text={fullUnit} />
+        </>
+      );
+    } else {
+      unitDisplay = <div className={cx('unit')}>{unit}</div>;
+    }
+  } else {
+    unitDisplay = null;
+  }
```
Comment (sdadn) [inline packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx:109]: We can simply this if ladder by using a ternary operator a second evaluation where it is used (see my next comment)
```suggestion
    const unitDisplay = fullUnit
         ? <>
           <div className={cx('unit')} aria-hidden="true">
             {unit}
           </div>
           <VisuallyHiddenText text={fullUnit} />
         </>
       : <div className={cx('unit')}>{unit}</div>;
```

File: packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx [hunk 4]
```diff
@@ -59,16 +70,51 @@ const ResultNameHeaderCell = (props) => {
...
       {...customProps}
       className={nameHeaderCellClassnames}
     >
-      <div className={cx('name')}>
-        {resultRowIndicators[typeIndicator.toLowerCase()]}
-        {resultName}
-      </div>
-      {unit && <div className={cx('unit')}>{unit}</div> }
+      {resultNameDisplay}
+      {unitDisplay}
```
Comment (sdadn) [inline packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx:117]:
```suggestion
      {unit && unitDisplay}
```

File: packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx [hunk 5]
```diff
@@ -13,6 +13,16 @@ describe('ResultNameHeaderCell', () => {
     expect(cell).toMatchSnapshot();
   });
 
+  it('should render correctly when a full result name is provided', () => {
+    const cell = shallow(<ResultNameHeaderCell resultName="Name" fullResultName="Full result name" />);
+    expect(cell).toMatchSnapshot();
```
Comment (sdadn) [inline packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx:18]: we should also explicitly check to see if the name is properly set
```suggestion
    expect(cell).toMatchSnapshot();
```

File: packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx [hunk 6]
```diff
@@ -13,6 +13,16 @@ describe('ResultNameHeaderCell', () => {
...
+  });
+
+  it('should render correctly when a full unit name is provided', () => {
+    const cell = shallow(<ResultNameHeaderCell resultName="Name" fullResultName="Full result name" unit="Unit" fullUnit="Full unit name" />);
+    expect(cell).toMatchSnapshot();
```
Comment (sdadn) [inline packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx:23]: same as above

**Review by sdadn — CHANGES_REQUESTED** [review]

File: packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx (hunk 3 above)
Comment (chenyixuan625) [reply packages/terra-clinical-result/src/result-name-header-cell/ResultNameHeaderCell.jsx]: Updated: https://github.com/cerner/terra-clinical/pull/867/commits/361d8279f67328213f80517a2e1ad39a1400c5cf#diff-10a8f15ae7fe78637c3fb3926670dd1d257ecc9c7a37f10b722d3ba6d579ea04R90-R99

File: packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx (hunk 5 above)
Comment (chenyixuan625) [reply packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx:19]: Good catch. Fixed: https://github.com/cerner/terra-clinical/pull/867/commits/361d8279f67328213f80517a2e1ad39a1400c5cf#diff-7d79d57b26e94be506526eda6c255971437300a335ab6c307c2f721a2a12b6f4R18

File: packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx (hunk 6 above)
Comment (chenyixuan625) [reply packages/terra-clinical-result/tests/jest/result-name-header-cell/ResultNameHeaderCell.test.jsx:25]: https://github.com/cerner/terra-clinical/pull/867/commits/361d8279f67328213f80517a2e1ad39a1400c5cf#diff-7d79d57b26e94be506526eda6c255971437300a335ab6c307c2f721a2a12b6f4R24

**Pull request merged**